util.go

```diff
@@ -50,19 +53,180 @@ func parseTag(path string) string {
+	replaceMap := make(map[string]string)
+	for s.Scan() {
+		ln := sanitizeLine(s.Text(), "//")
+		if ln == "" {
+			continue
+		}
+		parts := strings.Fields(ln)
+		numParts := len(parts)
+		if numParts != 4 {
+			if numParts == 1 && parts[0] == ")" {
+				// this is the end of the requires section, break out to process the others
+				break
+			}
+			return nil, fmt.Errorf("invalid config format: %s", ln)
+		}
+
+		commitOrVersion := getCommitOrVersion(parts[3])
+		if commitOrVersion == "" {
+			return nil, fmt.Errorf("invalid go.mod file, poorly formatted version in replace section %s", parts[3])
+		}
+		replaceMap[parts[0]] = commitOrVersion
+	}
+	if err := s.Err(); err != nil {
+		return nil, err
+	}
+	return replaceMap, nil
+}
+
+func processRequiresSection(s *bufio.Scanner) (map[string]*dependency, error) {
+	depMap := make(map[string]*dependency)
+	for s.Scan() {
+		ln := sanitizeLine(s.Text(), "//")
+		if ln == "" {
+			continue
+		}
+		parts := strings.Fields(ln)
+		numParts := len(parts)
+
+		if numParts != 2 {
+			if numParts == 1 && parts[0] == ")" {
+				// this is the end of the requires section, break out to process the others
+				break
+			}
+			return nil, fmt.Errorf("invalid config format: %s", ln)
+		}
+
+		commitOrVersion := getCommitOrVersion(parts[1])
+		if commitOrVersion == "" {
+			return nil, fmt.Errorf("invalid go.mod file, poorly formatted version in require section %s", parts[1])
+		}
+
+		depMap[parts[0]] = &dependency{
+			Name:     parts[0],
+			Commit:   commitOrVersion,
+			CloneURL: "git://" + parts[0],
+		}
+	}
+	if err := s.Err(); err != nil {
+		return nil, err
+	}
+	return depMap, nil
+}
+
+func sanitizeLine(line, commentDelim string) string {
+	ln := strings.TrimSpace(line)
+	if ln == "" {
+		return ""
+	}
+	cidx := strings.Index(ln, commentDelim)
+	// whole line is commented
+	if cidx == 0 {
+		return ""
+	}
+	if cidx > 0 {
+		ln = ln[:cidx]
+	}
+
+	return strings.TrimSpace(ln)
+}
+
+func getCommitOrVersion(cov string) string {
+	// parse the commit or version. It'll either be of the form
+	// v0.0.0 or v0.0.0-date-commitID. Split by '-' to check
+	dashFields := strings.FieldsFunc(cov, func(c rune) bool { return c == '-' })
+
+	if len(dashFields) == 1 || len(dashFields) == 2 {
+		// if dashFields came up empty, but the second parsed piece is not, it should be
+		// a version. use it
+		// it could also be a version with a '-' like v1.0.0-rc1
+		// in which case it should also be kept
+
+		// despite it being idiomatic to go modules, the +incompatible is a bit
+		// unsightly in release notes. Let's cut it out of the version if it
+		// exists
+		if incpIdx := strings.Index(cov, "+incompatible"); incpIdx > 0 {
+			return cov[:incpIdx]
+		}
+		return cov
+	} else if len(dashFields) == 3 {
+		// If there are three fields, use the last (the commit)
+		// as often the version found in the first field is just a placeholder
+		return dashFields[2]
```

Comment:
Maybe save the `+incompatible` stripping until the end, I saw this in a `go.mod`

`github.com/docker/distribution v2.7.1-0.20190205005809-0d3efadf0154+incompatible`